### PR TITLE
Add PBCs to the `System` class

### PR DESCRIPTION
**Pull request by frostedoyster**
This PR implements PBCs in the `System` class, allowing for structures where only some of the dimensions have PBCs.

A few questions:
- should we have default arguments (most likely `True, True, True` for backward compatibility)?
- is the strategy to ignore the cell if the PBCs are False? Or do we require it to have some all-zero rows if PBCs are False?

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [x] CHANGELOG updated with public API or any other important changes?

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/metatensor/metatensor/actions/artifacts/2004641647.zip)

### CODE REVIEW

**Review by Luthaf — COMMENTED** [review]
> should we have default arguments (most likely True, True, True for backward compatibility)?

With the same reasoning as in #646, I would not use a default value but rather force the user to think explicitly about this.

> is the strategy to ignore the cell if the PBCs are False? Or do we require it to have some all-zero rows if PBCs are False?

I'm not sure. I'd say it would be good to force the cell row to be 0 is the pbc is disabled along this dimension. If nothing else, this should help catch broken NL implementations downstream.

File: metatensor-torch/src/atomistic/system.cpp [hunk 1]
```diff
@@ -513,6 +546,38 @@ void SystemHolder::set_cell(torch::Tensor cell) {
     this->cell_ = std::move(cell);
 }
 
+void SystemHolder::set_pbc(torch::Tensor pbc) {
+    if (pbc.device() != this->device()) {
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/system.cpp:550]: should we move the constructor checks in this function & then call this function from constructor to perform the same checks? Right now this is missing the check for cell vectors vs pbc

File: metatensor-torch/src/atomistic/system.cpp [hunk 2]
```diff
@@ -410,6 +411,38 @@ SystemHolder::SystemHolder(torch::Tensor types, torch::Tensor positions, torch::
...
+    if (pbc_.scalar_type() != torch::kBool) {
+        C10_THROW_ERROR(ValueError,
+            "`pbc` must be a tensor of booleans, got " +
+            scalar_type_name(pbc_.scalar_type()) + " instead"
+        );
+    }
+
+    // if the pbcs are False along any directions, we check that the
+    // corresponding cell vectors are zero
+    if (pbc_.device().str() != "meta") {  // skip this check for meta tensors
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/system.cpp:438]:
```suggestion
    if (pbc_.device().is_meta()) {  // skip this check for meta tensors
```

**Review by Luthaf — COMMENTED** [review]
Looks good, could you update the CHANGELOG as well?

Also, this will need changes in the LAMMPS and PLUMED interfaces once we release the corresponding code, will you have time to do these changes? 

I think this is different enough from the previous code assumptions that it is worth doing a 0.6 (breaking release).

File: metatensor-torch/src/atomistic/system.cpp (hunk 1 above)
Comment (frostedoyster) [reply metatensor-torch/src/atomistic/system.cpp:550]: Moving all the checks into the setter meant dropping a lot of detail in the error messages. I've re-written the check in the setter so the error messages can be better

**Comment by frostedoyster** [comment]
I'd be happy do adapt the LAMMPS and PLUMED interfaces

File: metatensor-torch/CHANGELOG.md [hunk 3]
```diff
@@ -6,18 +6,19 @@ a changelog](https://keepachangelog.com/en/1.1.0/) format. This project follows
 
 ## [Unreleased](https://github.com/metatensor/metatensor/)
 
-<!-- Possible sections:
```
Comment (Luthaf) [inline metatensor-torch/CHANGELOG.md:9]: please leave this in place & copy the section you need below

File: metatensor-torch/src/atomistic/system.cpp (hunk 2 above)
Comment (Luthaf) [reply metatensor-torch/src/atomistic/system.cpp:438]: still relevant?

File: metatensor-torch/src/atomistic/system.cpp [hunk 4]
```diff
@@ -513,6 +546,49 @@ void SystemHolder::set_cell(torch::Tensor cell) {
...
+    if (pbc.size(0) != 3) {
+        C10_THROW_ERROR(ValueError,
+            "new `pbc` must contain 3 entries, got a tensor with " +
+            std::to_string(pbc.size(0)) + " values"
+        );
+    }
+
+    // if the pbcs are False along any directions, we check that the
+    // corresponding cell vectors are zero
+    if (pbc.device().str() != "meta") {  // skip this check for meta tensors
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/system.cpp:580]:
```suggestion
    if (pbc.device().is_meta()) {  // skip this check for meta tensors
```

Comment (frostedoyster) [reply metatensor-torch/src/atomistic/system.cpp:580]: Doesn't compile :(

Comment (Luthaf) [reply metatensor-torch/src/atomistic/system.cpp]: ok, I'm trying a comparison with `pbc_.device().type()` instead. I'll revert if it does not work

**Review by Luthaf — APPROVED** [review]

**Pull request merged**